In [1]:
import pandas as pd
df = pd.read_csv('서울특별시 공공자전거 이용정보 데이터.csv',engine="python",encoding = 'cp949')

In [2]:
df.tail(3)

대여일자  대여시간  대여소번호                 대여소명 대여구분코드 성별 연령대코드  이용건수  \
3704325  2022-06-30    23   4564          4564. 63스퀘어    정기권  M    기타     1   
3704326  2022-06-30    23    263       263. 근로자회관 사거리    정기권  M    기타     1   
3704327  2022-06-30    23   1308  1308. 안암로터리 버스정류장 앞    정기권  M    기타     1   

           운동량   탄소량  이동거리(M)  이용시간(분)  
3704325  65.64  0.51   2210.0       15  
3704326  38.81  0.23    980.0       16  
3704327  18.53  0.17    720.0        3

In [3]:
df.shape

(3704328, 12)

In [4]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3704328 entries, 0 to 3704327
Data columns (total 12 columns):
대여일자       object
대여시간       int64
대여소번호      int64
대여소명       object
대여구분코드     object
성별         object
연령대코드      object
이용건수       int64
운동량        object
탄소량        object
이동거리(M)    float64
이용시간(분)    int64
dtypes: float64(1), int64(4), object(7)
memory usage: 339.1+ MB


In [5]:
df.isna().sum()

대여일자             0
대여시간             0
대여소번호            0
대여소명             0
대여구분코드           0
성별         1261720
연령대코드            0
이용건수             0
운동량              0
탄소량              0
이동거리(M)          0
이용시간(분)          0
dtype: int64

 ### * 성별 중에 34% 는 결측치 -> 성별에 대한 정보 알아내기는 쉽지 않음

In [6]:
df.describe()

대여시간         대여소번호          이용건수       이동거리(M)       이용시간(분)
count  3.704328e+06  3.704328e+06  3.704328e+06  3.704328e+06  3.704328e+06
mean   1.416412e+01  1.987047e+03  1.149120e+00  3.135646e+03  2.622830e+01
std    5.984212e+00  1.455010e+03  4.984830e-01  3.944730e+03  3.192456e+01
min    0.000000e+00  3.000000e+00  1.000000e+00  0.000000e+00  0.000000e+00
25%    1.000000e+01  7.740000e+02  1.000000e+00  9.700000e+02  7.000000e+00
50%    1.600000e+01  1.628000e+03  1.000000e+00  1.800810e+03  1.500000e+01
75%    1.900000e+01  3.006000e+03  1.000000e+00  3.630000e+03  3.400000e+01
max    2.300000e+01  9.999900e+04  3.100000e+01  1.952100e+05  1.938000e+03

In [7]:
df['대여시간'].value_counts()

18    335855
17    278332
19    265785
20    231713
8     226663
16    220779
21    216761
22    184793
15    179002
7     167462
12    158375
13    155557
14    153363
9     149161
11    146412
10    127809
23    127080
0     102343
6      73402
1      69609
2      45710
5      36822
3      28577
4      22963
Name: 대여시간, dtype: int64

In [8]:
df['연령대코드'].value_counts()

20대      1257122
30대       862500
40대       523899
기타        376098
50대       338900
~10대      227169
60대       103920
70대이상      14720
Name: 연령대코드, dtype: int64

In [9]:
print(1257122/3704328)
print(862500/3704328)
print(523899/3704328)

0.33936573651145363
0.2328357532054397
0.14142889074617582


### * 20대가 33%, 30대가 23%, 40대가 14% => 총 70%
### * 기타 => 10% // 6~70대 3% // ~10대 6%

In [10]:
df['이용건수'].value_counts()

1     3282763
2      336329
3       59719
4       15903
5        5235
6        2120
7         978
8         471
9         295
10        156
11        122
12         68
13         36
15         36
14         31
16         19
17         14
18         13
20          6
19          4
21          3
23          3
22          2
24          1
31          1
Name: 이용건수, dtype: int64

### * 한달 기준 이용횟수 1~ 6번이 제일 많음

In [11]:
df = df.rename(columns = {'이용시간(분)' : '이용시간'})
df = df.rename(columns = {'이동거리(M)': '이동거리'})
df.head(3)

대여일자  대여시간  대여소번호                대여소명 대여구분코드   성별 연령대코드  이용건수    운동량  \
0  2022-06-01     0    742    742. 등촌역 5번 출구 뒤    정기권  NaN  ~10대     1  48.30   
1  2022-06-01     0   2059      2059. 보라매공원 정문    정기권  NaN  ~10대     1  27.47   
2  2022-06-01     0   1349  1349. 월계2교 버스정류장 앞    정기권  NaN  ~10대     1  20.63   

    탄소량     이동거리  이용시간  
0  0.46  1999.70    10  
1  0.23   991.00     9  
2  0.20   868.17     3

In [12]:
df['이용시간'].value_counts()

5       183379
6       179361
4       174703
7       168394
8       156871
9       145901
3       141923
10      135563
11      122572
12      112678
13      102287
14       93800
2        89257
15       86298
16       79441
17       72513
18       67657
19       62323
20       59310
21       54488
22       51237
23       47858
24       45320
25       42695
26       40508
27       37995
28       36166
29       34351
30       33711
31       31670
         ...  
778          1
779          1
783          1
790          1
791          1
792          1
793          1
509          1
751          1
749          1
748          1
747          1
716          1
717          1
718          1
719          1
528          1
721          1
722          1
725          1
726          1
730          1
732          1
733          1
735          1
526          1
741          1
742          1
524          1
1938         1
Name: 이용시간, Length: 817, dtype: int64

In [13]:
df['이용시간'].value_counts().head(15)

5     183379
6     179361
4     174703
7     168394
8     156871
9     145901
3     141923
10    135563
11    122572
12    112678
13    102287
14     93800
2      89257
15     86298
16     79441
Name: 이용시간, dtype: int64

In [14]:
sum(df['이용시간']<= 30)

2687055

In [15]:
sum(df['이용시간']<= 15)

1921482

In [16]:
print(2687055/3704328)

0.7253825795123974


In [17]:
print(1921482/3704328)

0.5187127057863127


### * 이용자의 72.5%는 30분 이내로 자전거를 대여함
### * 이용자의 51.8%는 15분 이내로 자전거를 대여함

In [18]:
go_time = df[(df['대여시간']<=10)&(df['대여시간']>=7)]['이용시간']

In [19]:
back_time = df[(df['대여시간']<=20)&(df['대여시간']>=17)]['이용시간']

In [20]:
go_time.value_counts().head(15)

4     44980
5     44184
6     40699
3     37978
7     36753
8     33413
9     30387
10    27581
2     24706
11    24643
12    22188
13    19387
14    17502
15    16046
16    14472
Name: 이용시간, dtype: int64

In [21]:
sum(go_time <= 15)

428802

In [22]:
sum(go_time <= 30)

556302

In [23]:
print(428802/3704328)

0.1157570279953611


In [24]:
print(556302/3704328)

0.15017622629529567


In [25]:
back_time.value_counts().head(15)

6     45864
5     45225
7     44055
4     41754
8     41482
9     39350
10    37006
11    33937
3     32294
12    31738
13    29411
14    27300
15    25327
16    23515
17    21809
Name: 이용시간, dtype: int64

In [26]:
sum(back_time <= 15)

500515

In [27]:
sum(back_time <= 30)

741542

In [28]:
print(500515/3704328)

0.13511627480071958


In [29]:
print(741542/3704328)

0.20018259722141235


### - 출근시간(7~10)에 15분 이하 이용자 11.7%
### - 퇴근시간(17~20)에 15분 이하 이용자 13.5%    => 총 25.2%

# --------------------------------------------

### - 출근시간(7~10)에 30분 이하 이용자 15.0%
### - 퇴근시간(17~20)에 30분 이하 이용자 20.0%    => 총 35.0%

In [30]:
# 경고메세지 제거
import warnings
warnings.filterwarnings('ignore')
import matplotlib.pyplot as plt
from matplotlib import font_manager

font_fname = 'c:/windows/fonts/malgun.ttf'
font_family = font_manager.FontProperties(fname=font_fname).get_name()
plt.rcParams['font.family'] = font_family

import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [32]:
df_time = pd.DataFrame(df['연령대코드'].groupby(df['대여시간']).value_counts())
df_time.columns = ['cnt']
df_time = df_time.reset_index()

In [33]:
df_age = pd.DataFrame(df['연령대코드'].groupby(df['대여시간']).value_counts())
df_age.columns = ['cnt']
df_age = df_age.reset_index()

In [34]:
df_age = pd.pivot_table(df_age, index = '대여시간', columns = '연령대코드', values = 'cnt')
df_age = df_age.reset_index()
df_age['청년이용률'] = round((df_age['20대']+df_age['30대']+df_age['40대'])/
                        (df_age['20대']+df_age['30대']+df_age['40대']+df_age['50대']+df_age['60대']
                        +df_age['70대이상']+df_age['~10대']+df_age['기타'])*100, 1)
# df_age = df_age.sort_values(by = ['청년이용률'], ascending = False)

In [35]:
df_age

연령대코드  대여시간     20대    30대    40대    50대   60대  70대이상   ~10대     기타  청년이용률
0         0   43777  18953   9057   7235  2624    266   9306  11125   70.1
1         1   27970  12004   6463   6773  2818    275   5685   7621   66.7
2         2   18312   7258   4196   5134  2311    198   3316   4985   65.1
3         3   10592   4320   3024   3864  1783    150   1722   3122   62.8
4         4    7633   3681   2917   3384  1551    125   1185   2487   62.0
5         5   10173   7595   5980   5574  2571    338   1013   3578   64.5
6         6   18317  17420  14161  10651  3688    529   2056   6580   68.0
7         7   40451  39090  29927  20157  5410    643  15024  16760   65.4
8         8   66723  61919  40945  23112  5684    914   6340  21026   74.8
9         9   48344  37692  23580  15291  4915    806   4515  14018   73.5
10       10   42233  30044  19737  13254  4895    862   4843  11941   72.0
11       11   48595  35612  22912  14227  4943    851   5747  13525   73.2
12       12   53674  38095  23735  15038  4785    815   7013  15220   72.9
13       13   54069  35852  22518  14767  4873    719   7745  15014   72.3
14       14   53653  34413  22130  14024  4833    827   7969  15514   71.9
15       15   61460  37940  25248  16603  5838    956  12045  18912   69.6
16       16   70677  46108  30812  20442  6874   1107  20053  24706   66.9
17       17   89095  64780  42132  26231  7854   1076  18276  28888   70.4
18       18  107778  85456  52309  30484  7355    944  16860  34669   73.1
19       19   89983  67592  37175  22015  6027    798  14953  27242   73.3
20       20   83039  57871  30277  17530  4336    549  14653  23458   73.9
21       21   81772  51912  25178  14956  3376    410  16226  22931   73.3
22       22   73734  40439  18129  10833  2585    326  18974  19773   71.6
23       23   55068  26454  11357   7321  1991    236  11650  13003   73.1